src/components/VariantSelector/VariantSelector.ts

```diff
@@ -0,0 +1,181 @@
+
+    // Initialize selections with first value of each option
+    initializeDefaultSelections(element, productData)
+
+    const selectorHTML = generateVariantSelectorHTML(element, productData)
+
+    // Use constructible stylesheets if supported, fallback to inline styles
+    if ("adoptedStyleSheets" in element.shadowRoot!) {
+      if (!cachedStyleSheet) {
+        cachedStyleSheet = new CSSStyleSheet()
+        await cachedStyleSheet.replace(variantSelectorStyles)
+      }
+      element.shadowRoot!.adoptedStyleSheets = [cachedStyleSheet]
+      element.shadowRoot!.innerHTML = selectorHTML.html
+    } else {
+      element.shadowRoot!.innerHTML = `
+        <style>${variantSelectorStyles}</style>
+        ${selectorHTML.html}
+      `
+    }
+
+    // Setup event listeners for option buttons
+    setupOptionListeners(element)
+
+    updateActiveStates(element)
+    emitVariantChange(element, productData)
+  } finally {
+    element.toggleAttribute("loading", false)
+  }
+}
+
+function initializeDefaultSelections(element: VariantSelector, product: ShopifyProduct) {
+  product.options.forEach(option => {
+    if (option.values.length > 0) {
+      element.selectedOptions[option.name] = option.values[0]
+    }
+  })
+}
+
+function setupOptionListeners(element: VariantSelector) {
+  if (!element.shadowRoot) return
+
+  element.shadowRoot.addEventListener("click", async e => {
+    const target = e.target as HTMLElement
+    if (target.classList.contains("variant-option-value")) {
+      e.preventDefault()
+      const optionName = target.getAttribute("data-option-name")
+      const optionValue = target.getAttribute("data-option-value")
+
+      if (optionName && optionValue) {
+        await selectOption(element, optionName, optionValue)
+      }
+    }
+  })
+}
+
+export async function selectOption(element: VariantSelector, optionName: string, value: string) {
+  element.selectedOptions[optionName] = value
+  updateActiveStates(element)
+
+  // Fetch product data and emit variant change
+  const productData = await fetchProductData(element.handle)
+  emitVariantChange(element, productData)
+}
+
+function updateActiveStates(element: VariantSelector) {
+  if (!element.shadowRoot) return
+
+  element.shadowRoot.querySelectorAll(".variant-option-value").forEach(button => {
+    const optionName = button.getAttribute("data-option-name")
+    const optionValue = button.getAttribute("data-option-value")
+
+    if (optionName && element.selectedOptions[optionName] === optionValue) {
+      button.classList.add("variant-option-value--active")
+    } else {
+      button.classList.remove("variant-option-value--active")
+    }
+  })
+}
+
+function emitVariantChange(element: VariantSelector, product: ShopifyProduct) {
+  const variant = getSelectedVariant(element, product)
+  if (variant) {
+    const detail: VariantChangeDetail = { variant }
+    element.dispatchEvent(
+      new CustomEvent("variantchange", {
```

Comment:
for future PRs, consider a strictly typed event handling like bus-event? It will get easier to document and maintain all these events across components. Something like event.ts for products

Reply:
good idea